pennylane/noise/conditionals.py

```diff
@@ -0,0 +1,462 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains utility functions for building boolean conditionals for noise models"""
+
+from functools import partial
+from inspect import isclass, signature
+
+import pennylane as qml
+from pennylane.boolean_fn import BooleanFn
+from pennylane.wires import Wires
+
+# pylint: disable = unnecessary-lambda, too-few-public-methods
+
+
+class NoiseConditional(BooleanFn):
+    """Wrapper for callables with boolean output that help implement noise models
+    and can be manipulated via bit-wise operations.
+
+    Args:
+        fn (callable): Function to be wrapped. It can accept any number
+            of arguments, and must return a boolean.
+        repr (str): String representation to be used by ``repr`` dunder method.
+            Default is to use function's name as ``NoiseConditional(<name>)``.
+
+    .. note:: This is a *developer-facing* class for implementing conditionals for building
+        noise models. Users are encouraged to rather use :func:`~.wires_in`, :func:`~.wires_eq`,
+        :func:`~.op_in`, and :func:`~.op_eq` for this purpose.
+    """
+
+    def __init__(self, fn, repr=None):
+        super().__init__(fn)
+        self.repr = repr if repr else f"NoiseConditional({fn.__name__})"
+
+    def __and__(self, other):
+        return AndConditional(self, other)
+
+    def __or__(self, other):
+        return OrConditional(self, other)
+
+    def __repr__(self):
+        return self.repr
+
+
+# pylint: disable = bad-super-call
+class AndConditional(NoiseConditional):
+    """Developer facing class for implemeting bit-wise ``AND`` for callables
+    wrapped up with :class:`NoiseConditional <pennylane.noise.NoiseConditional>`.
+
+    Args:
+        left (~.NoiseConditional): Left operand in the bit-wise expression.
+        right (~.NoiseConditional): Right operand in the bit-wise expression.
+
+    .. note::
+
+        This is a *developer-facing* class for implementing bit-wise expression
+        for the noise model's conditionals built via
+        :class:`NoiseConditional <pennylane.noise.NoiseConditional>`.
```

Comment:
Nice! My only suggestion might be to not make it sound too developer facing, as users will actually have access to this object even if they don't directly instantiate it.